File: src/php/src/MergeSort.php
<?php

// IMPLEMENTAÇÃO MERGE SORT EM PHP
class MergeSort {
    private $array = [];
    private $arqE;
    private $arqS;

    public function __construct($arqE, $arqS) {
        $this->arq = $arqE;
        $this->arq2 = $arqS;
    }

    private function merge(&$v, $inicio, $meio, $fim) {
        $temp = [];
        $p1 = $inicio;
        $p2 = $meio + 1;
        $i = 0;

        while ($p1 <= $meio && $p2 <= $fim) {
            if ($v[$p1] < $v[$p2]) {
                $temp[$i++] = $v[$p1++];
            } else {
                $temp[$i++] = $v[$p2++];
            }
        }

        while ($p1 <= $meio) {
            $temp[$i++] = $v[$p1++];
        }

        while ($p2 <= $fim) {
            $temp[$i++] = $v[$p2++];
        }

        for ($j = 0, $k = $inicio; $j < count($temp); $j++, $k++) {
            $v[$k] = $temp[$j];
        }
    }

    private function mergeSort(&$v, $inicio, $fim) {
        if ($inicio < $fim) {
            $meio = (int) ($inicio + ($fim - $inicio) / 2);
            $this->mergeSort($v, $inicio, $meio);
            $this->mergeSort($v, $meio + 1, $fim);
            $this->merge($v, $inicio, $meio, $fim);
        }
    }

    private function definirArray($n) {
        $content = file_get_contents($this->arq);
        $numbers = explode(' ', $content);

        $this->array = [];

        foreach ($numbers as $num) {
            $num = trim($num);
            if ($num !== '') {
                $this->array[] = (int) $num;
            }
        }

        $this->array = array_slice($this->array, 0, $n);
    }

    private function salvarTempo($n, $time_taken) {
        $data = "\nPHP,$n," . number_format($time_taken, 10) . ",$this->arq";
        file_put_contents($this->arq2, $data, FILE_APPEND);
        echo "Tempo de execução:" . number_format($time_taken, 10);
    }

    public function run($n) {
        $this->definirArray($n);

        $start = microtime(true);
        $this->mergeSort($this->array, 0, count($this->array) - 1);
        $end = microtime(true);
        $time_taken = ($end - $start);

        $this->salvarTempo($n, $time_taken);
    }
}
?>

File: src/rust/src/merge_sort.rs
use std::fs::{File, OpenOptions};
use std::io::{BufRead, BufReader, Write};
use std::time::Instant;

pub struct MergeSort {
    array: Vec<i32>,
    arq: String,
    arq2: String,
}

impl MergeSort {
    pub fn new(arq: String, arq2: String) -> MergeSort {
        MergeSort {
            array: Vec::new(),
            arq,
            arq2,
        }
    }

    fn merge(&mut self, v: &mut [i32], inicio: usize, meio: usize, fim: usize) {
        let mut temp = Vec::with_capacity(fim - inicio + 1);
        let (mut p1, mut p2) = (inicio, meio + 1);

        while p1 <= meio && p2 <= fim {
            if v[p1] <= v[p2] {
                temp.push(v[p1]);
                p1 += 1;
            } else {
                temp.push(v[p2]);
                p2 += 1;
            }
        }

        while p1 <= meio {
            temp.push(v[p1]);
            p1 += 1;
        }

        while p2 <= fim {
            temp.push(v[p2]);
            p2 += 1;
        }

        for (i, val) in temp.iter().enumerate() {
            v[inicio + i] = *val;
        }
    }

    fn merge_sort_helper(&mut self, v: &mut [i32], inicio: usize, fim: usize) {
        if inicio < fim {
            let meio = (inicio + fim) / 2;
            let (left, right) = v.split_at_mut(meio + 1);
            self.merge_sort_helper(left, inicio, meio);
            self.merge_sort_helper(right, 0, fim - meio - 1);
            self.merge(v, inicio, meio, fim);
        }
    }

    pub fn run(&mut self, n: usize) {
        self.definir_array(n);

        let start = Instant::now();
        let len = self.array.len();
        let mut array_copy = self.array.clone(); // Clonar o vetor para evitar mutabilidade simultânea
        self.merge_sort_helper(&mut array_copy, 0, len - 1);
        let duration = start.elapsed();

        self.salvar_tempo(n, duration.as_secs_f64());
    }

    fn definir_array(&mut self, n: usize) {
        let file = File::open(&self.arq).expect("Erro ao abrir o arquivo de entrada");
        let reader = BufReader::new(file);
        
        self.array.clear();
    
        // Lê a primeira linha do arquivo
        if let Some(Ok(line)) = reader.lines().next() {
            // Divide a linha em números separados por espaço
            for num_str in line.split_whitespace() {
                if self.array.len() >= n {
                    break;
                }
                if let Ok(num) = num_str.parse::<i32>() {
                    self.array.push(num);
                } else {
                    println!("Número inválido: {}", num_str);
                }
            }
        }
    }

    fn salvar_tempo(&self, n: usize, time_taken: f64) {
        let mut file = OpenOptions::new()
            .write(true)
            .append(true)
            .open(&self.arq2)
            .expect("Erro ao abrir o arquivo de saída");

        write!(file, "\nRust, {}, {:.10}, {}", n, time_taken, self.arq).expect("Erro ao escrever no arquivo");

        println!("\nTempo de execução: {:.10}", time_taken);
    }
}
